roxie/roxiemem/roxiemem.cpp

```diff
@@ -1099,6 +1106,117 @@ bool HeapletBase::hasDestructor(const void *ptr)
+const static size_t parallelSyncReleaseThreshold = DEFAULT_PARALLEL_SYNC_RELEASE_THRESHOLD;
+#endif
+
+
+inline void inlineReleaseRoxieRowArray(size_t count, const void * * rows)
+{
+    for (size_t i = 0; i < count; i++)
+        ReleaseRoxieRow(rows[i]);
+}
+
+#ifdef PARALLEL_SYNC_RELEASE
+class sync_releaser_task : public tbb::task
+{
+public:
+    sync_releaser_task(size_t _num, const void * * _rows) : num(_num), rows(_rows) {}
+
+    virtual task * execute()
+    {
+        inlineReleaseRoxieRowArray(num, rows);
+        return NULL;
+    }
+
+public:
+    size_t num;
+    const void * * rows;
+};
+
+void ParallelReleaseRoxieRowArray(size_t count, const void * * rows)
+{
+    //This is quicker than creating a task list and then spawning all.
+    tbb::task * completed_task = new (tbb::task::allocate_root()) tbb::empty_task();
+    unsigned numTasks = (unsigned)((count + parallelSyncReleaseGranularity -1) / parallelSyncReleaseGranularity);
+    completed_task->set_ref_count(1+numTasks);
+
+    //Arrange the blocks that are freed from the array so that if the rows are allocated in order, adjacent blocks
+    //are not released at the same time, therefore reducing contention.
+    const unsigned stride = 8;
+    for (unsigned j= 0; j < stride; j++)
+    {
+        size_t i = j * parallelSyncReleaseGranularity;
+        for (; i < count; i += parallelSyncReleaseGranularity * stride)
+        {
+            size_t remain = count - i;
+            size_t blockRows = (remain > parallelSyncReleaseGranularity) ? parallelSyncReleaseGranularity : remain;
+            tbb::task * next = new (completed_task->allocate_child()) sync_releaser_task(blockRows, rows + i);
+            next->spawn(*next); // static member in tbb 3.0
+        }
+    }
+
+    completed_task->wait_for_all();
+    completed_task->destroy(*completed_task);       // static member in tbb 3.0
+}
+#endif
+
+
+void ReleaseRoxieRowArray(size_t count, const void * * rows)
+{
+#ifdef PARALLEL_SYNC_RELEASE
+    if (count >= parallelSyncReleaseThreshold)
+    {
+        ParallelReleaseRoxieRowArray(count, rows);
+        return;
+    }
+#endif
+
+    inlineReleaseRoxieRowArray(count, rows);
+}
+
+//Not implemented as ReleaseRoxieRowArray(to - from, rows + from) to avoid "from > to" wrapping issues
+void roxiemem_decl ReleaseRoxieRowRange(const void * * rows, size_t from, size_t to)
+{
+#ifdef PARALLEL_SYNC_RELEASE
+    if ((from < to) && ((to - from) >= parallelSyncReleaseThreshold))
+    {
+        ParallelReleaseRoxieRowArray(to-from, rows+from);
+        return;
+    }
+#endif
+    for (size_t i = from; i < to; i++)
+        ReleaseRoxieRow(rows[i]);
+}
+
+
+void ReleaseRoxieRows(ConstPointerArray &rows)
+{
+    ReleaseRoxieRowArray(rows.ordinality(), rows.getArray());
+    rows.kill();
+}
+
+void ReleaseRoxieRows(ConstPointerArray & rows, size_t first)
+{
+    ReleaseRoxieRowRange(rows.getArray(), first, rows.ordinality());
+    rows.kill();
```

Comment:
Either that or have people call ReleaseRoxieRowArray and explicitly do the kill() themselves.